tests/components/xiaomi_miio/test_select.py

```diff
@@ -0,0 +1,267 @@
+"""The tests for the xiaomi_miio select component."""
+
+import enum
+import logging
+from unittest.mock import patch
+
+from arrow import utcnow
+import pytest
+
+from homeassistant.components.asuswrt.router import KEY_COORDINATOR
+from homeassistant.components.select import DOMAIN
+from homeassistant.components.select.const import (
+    ATTR_OPTION,
+    ATTR_OPTIONS,
+    SERVICE_SELECT_OPTION,
+)
+from homeassistant.components.xiaomi_miio import (
+    UPDATE_INTERVAL,
+    _async_update_data_default,
+)
+from homeassistant.components.xiaomi_miio.const import (
+    CONF_DEVICE,
+    CONF_FLOW_TYPE,
+    CONF_MAC,
+    DOMAIN as XIAOMI_DOMAIN,
+    KEY_DEVICE,
+)
+from homeassistant.components.xiaomi_miio.select import (
+    AttributeEnumMapping,
+    XiaomiMiioSelectDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    CONF_HOST,
+    CONF_MODEL,
+    CONF_TOKEN,
+    Platform,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from tests.common import MockConfigEntry, async_fire_time_changed
+from tests.components.xiaomi_miio import TEST_MAC
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DummyEnum_0(enum.Enum):
+    """Enum to test auto naming."""
+
+    Option_0 = 0
+    Option_1 = 1
+
+
+class DummyEnum_1(enum.Enum):
+    """Enum to test predefined naming."""
+
+    BadName_0 = 0
+    BadName_1 = 1
+
+
+class DummySelectStatus:
+    """Dummy device status."""
+
+    def __init__(self):
+        """Initialize the dummy device status."""
+        self.dummy = DummyEnum_0.Option_1
+        self.bad_name = DummyEnum_1.BadName_1
+
+
+class DummySelectDevice:
+    """Dummy device."""
+
+    def __init__(self, model, *args, **kwargs):
+        """Initialize the dummy device."""
+        self.data = DummySelectStatus()
+
+    def status(self):
+        """Retrieve properties."""
+        return self.data
+
+    def set_dummy_value(self, value: DummyEnum_0):
+        """Set dummy enum value."""
+        self.data.dummy = value
+        return True
+
+    def set_bad_name_value(self, value: DummyEnum_1):
+        """Set bad name enum value."""
+        self.data.bad_name = value
+        return True
+
+
+TEST_MODEL_TO_ATTR_MAP: dict[str, list] = {
+    "Dummy": [
+        AttributeEnumMapping("dummy", DummyEnum_0),
+        AttributeEnumMapping("bad_name", DummyEnum_1),
+    ]
+}
+
+TEST_SELECTOR_TYPES = (
+    XiaomiMiioSelectDescription(
```

Comment:
We shouldn't define any entity descriptions in the test. We have already defined entity descriptions in the select platform. We should just set up the integration while patching the library as needed and let the integration set up the select platform and add an entity. Then assert the entity state in the state machine and call services via the service registry.

Reply:
:+1: We want to have the complete code path from description definitions to initializing the entities to calling the defined setters tested. I think it would be fine to make the tests against a single select on a single model to verify that the code path works throughout the chain.

The idea of using a dummy device instead of targeting a specific implementation for testing came from me, as I'm currently working on making the backend library deliver the necessary information about available settings, sensors etc. to avoid hard-coding any of the details inside homeassistant. 

@MartinHjelmare If you have an idea what's the best approach (or even an example integration that does this properly) to test such dynamically created entities, it would be really helpful. The meta data delivered by the library will look something like this for selects:
```
@dataclass
class EnumSetting:
    name: str
    icon: str
    choices: Enum
    setter: Callable
```

Reply:
@MartinHjelmare Rewriting with using, one real device. Could you look please.